detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/DoubleNegativeLambda.kt

```diff
@@ -0,0 +1,130 @@
+            "that also contains negation. This is more readable when rewritten using a positive form of the function " +
+            "(like `takeIf`).",
+        Debt.FIVE_MINS,
+    )
+
+    private val splitCamelCaseRegex = "(?<=[a-z])(?=[A-Z])".toRegex()
+
+    private val negationTokens = listOf(
+        KtTokens.EXCL,
+        KtTokens.EXCLEQ,
+        KtTokens.EXCLEQEQEQ,
+        KtTokens.NOT_IN,
+        KtTokens.NOT_IS,
+    )
+
+    @Configuration(
+        "Function names expressed in the negative that can form double negatives with their lambda blocks. " +
+            "These are grouped together with the positive counterpart of the function, or `null` if this is unknown."
+    )
+    private val negativeFunctions: List<NegativeFunction> by config(
+        valuesWithReason(
+            "takeUnless" to "takeIf",
+        )
+    ) { list ->
+        list.map { NegativeFunction(simpleName = it.value, positiveCounterpart = it.reason) }
+    }
+
+    @Configuration(
+        "Function name parts to look for in the lambda block when deciding " +
+            "if the lambda contains a negative."
+    )
+    private val negativeFunctionNameParts: Set<String> by config(listOf("not", "non")) { it.toSet() }
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        super.visitCallExpression(expression)
+        val calleeExpression = expression.calleeExpression?.text ?: return
+        val negativeFunction = negativeFunctions.firstOrNull { it.simpleName == calleeExpression } ?: return
+        val lambdaArgument = expression.lambdaArguments.firstOrNull() ?: return
+
+        val forbiddenChildren = lambdaArgument.collectDescendantsOfType<KtExpression> {
+            it.isForbiddenNegation()
+        }
+
+        if (forbiddenChildren.isNotEmpty()) {
+            report(
+                CodeSmell(
+                    issue,
+                    Entity.from(expression),
+                    formatMessage(forbiddenChildren, negativeFunction)
+                )
+            )
+        }
+    }
+
+    private fun KtExpression.isForbiddenNegation(): Boolean {
+        return when (this) {
+            is KtOperationReferenceExpression -> operationSignTokenType in negationTokens
+            is KtCallExpression -> text == "not()" || text.split(splitCamelCaseRegex).map { it.lowercase() }
+                .any { it in negativeFunctionNameParts }
+
+            else -> false
+        }
+    }
+
+    private fun formatMessage(
+        forbiddenChildren: List<KtExpression>,
+        negativeFunction: NegativeFunction,
+    ) = buildString {
+        append("Double negative through using ${forbiddenChildren.joinInBackTicks()} inside a ")
+        append("`${negativeFunction.simpleName}` lambda. ")
+        append("Rewrite in the positive")
+        if (negativeFunction.positiveCounterpart != null) {
+            append(" with `${negativeFunction.positiveCounterpart}`.")
+        } else {
+            append(".")
+        }
```

Comment:
```suggestion
        if (negativeFunction.positiveCounterpart != null) {
            append(negativeFunction.positiveCounterpart")
        } else {
            append("Rewrite in the positive.")
        }
```

I think that `reason` should contain a reason, not just the counter part. And that should also make this code a bit simplier. Don't commit exactly my code, I assume that further rework is needed to land this change.

Reply:
Thanks, that makes a lot more sense. I did the change in 2ded6a41c 

I called it a "recommendation" ("Use `takeIf` instead") rather than a "reason" because that seems to be what it's called in the docs:
https://github.com/detekt/detekt/blob/main/.github/CONTRIBUTING.md

but I can rename it back to "reason" if you prefer.